server/src/main/java/org/opensearch/gateway/TransportNodesListGatewayStartedBatchShards.java

```diff
@@ -0,0 +1,460 @@
+        AsyncShardFetch.Lister<
+            TransportNodesListGatewayStartedShardsBatch.NodesGatewayStartedShardsBatch,
+            TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> {
+
+    public static final String ACTION_NAME = "internal:gateway/local/started_shards_batch";
+    public static final ActionType<NodesGatewayStartedShardsBatch> TYPE = new ActionType<>(
+        ACTION_NAME,
+        NodesGatewayStartedShardsBatch::new
+    );
+
+    private final Settings settings;
+    private final NodeEnvironment nodeEnv;
+    private final IndicesService indicesService;
+    private final NamedXContentRegistry namedXContentRegistry;
+
+    @Inject
+    public TransportNodesListGatewayStartedShardsBatch(
+        Settings settings,
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NodeEnvironment env,
+        IndicesService indicesService,
+        NamedXContentRegistry namedXContentRegistry
+    ) {
+        super(
+            ACTION_NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            Request::new,
+            NodeRequest::new,
+            ThreadPool.Names.FETCH_SHARD_STARTED,
+            NodeGatewayStartedShardsBatch.class
+        );
+        this.settings = settings;
+        this.nodeEnv = env;
+        this.indicesService = indicesService;
+        this.namedXContentRegistry = namedXContentRegistry;
+    }
+
+    @Override
+    public void list(
+        Map<ShardId, String> shardIdsWithCustomDataPath,
+        DiscoveryNode[] nodes,
+        ActionListener<NodesGatewayStartedShardsBatch> listener
+    ) {
+        execute(new Request(nodes, shardIdsWithCustomDataPath), listener);
+    }
+
+    @Override
+    protected NodeRequest newNodeRequest(Request request) {
+        return new NodeRequest(request);
+    }
+
+    @Override
+    protected NodeGatewayStartedShardsBatch newNodeResponse(StreamInput in) throws IOException {
+        return new NodeGatewayStartedShardsBatch(in);
+    }
+
+    @Override
+    protected NodesGatewayStartedShardsBatch newResponse(
+        Request request,
+        List<NodeGatewayStartedShardsBatch> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new NodesGatewayStartedShardsBatch(clusterService.getClusterName(), responses, failures);
+    }
+
+    /**
+     * This function is similar to nodeOperation method of {@link TransportNodesListGatewayStartedShards} we loop over
+     * the shards here and populate the data about the shards held by the local node.
+     *
+     * @param request Request containing the map shardIdsWithCustomDataPath.
+     * @return NodeGatewayStartedShardsBatch contains the data about the primary shards held by the local node
+     */
+    @Override
+    protected NodeGatewayStartedShardsBatch nodeOperation(NodeRequest request) {
```

Comment:
@Bukhtawar @shwetathareja There was a suggestion to remove helper class earlier but this method stays duplicated. Do you think it is ok or we should try reusing?

Reply:
can we compose batch operation leveraging individual operations to avoid dupliation?

Reply:
https://github.com/opensearch-project/OpenSearch/blob/ea48cb3ef3bba243de02b89d93cec17f4cf51d7f/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L151-L223

Plan is to keep only 1 method in helper class what I pasted above, and both current and new transport will use the helper class method.

@shwetathareja  - please confirm if creating that helper is okay ?

Reply:
@amkhar : I am fine moving to helper class but ensure it doesnt become dumping ground for random method and that class should get deprecated as soon as we clean up the old code of single shard fetch

Reply:
@vikasvb90 @shwetathareja pushed a commit with a Helper class that has a single method mentioned above.